File: solutions/Two Pointers/167_two-sum-ii-input-array-is-sorted.py
from typing import List

class Solution:
    def twoSum(self, numbers: List[int], target: int) -> List[int]:
        h = {}

        for ind, val in enumerate(numbers):
            if val in h:
                return [h[val], ind +1]
            else:
                h[target - val] = ind + 1

# Time Complexity: O(n)
# Space Complexity: O(n)

class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        l,r = 0, len(nums) -1
        
        while l < r:
            x = nums[l] + nums[r]
            if x == target:
                return [l +1, r +1]

            if x > target:
                r -= 1
            else:
                l += 1

# Time Complexity: O(n)
# Space Complexity: O(1)

File: C++/Two Pointers/167_two-sum-ii-input-array-is-sorted.cpp
#include <vector>
#include <unordered_map>

using namespace std;

class Solution {
public:
    vector<int> twoSum(vector<int>& numbers, int target) {
        int l = 0, r = numbers.size() - 1;

        while (l < r) {
            int sum = numbers[l] + numbers[r];

            if (sum == target) return {++l, ++r};
            else if (sum < target) l++;
            else r--;
        }

        return {0,0};
    }
};


class Solution {
public:
    vector<int> twoSum(vector<int>& numbers, int target) {
        size_t n = numbers.size();
        unordered_map<int, int> map;
        int num;

        for (int i = 0; i < n; ++i) {
            num = numbers[i];
            if (map.contains(num)) return {map[num], ++i};
            map[target - num] = i + 1;
        }

        return {0,0};
    }
};